src/plugins/intel_gpu/CMakeLists.txt

```diff
@@ -65,10 +65,14 @@ ov_add_plugin(NAME ${TARGET_NAME}
 target_compile_options(${TARGET_NAME} PRIVATE
     $<$<CONFIG:Release>:$<IF:$<CXX_COMPILER_ID:MSVC>,/Os,-Os>>)
 
-target_link_libraries(${TARGET_NAME} PRIVATE openvino_intel_gpu_graph openvino::pugixml)
+target_link_libraries(
+    ${TARGET_NAME} PRIVATE openvino_intel_gpu_graph openvino::pugixml ${GRAPH_COMPILER_LIBS})
 
 target_include_directories(${TARGET_NAME} PRIVATE
-        ${CMAKE_CURRENT_SOURCE_DIR}/include/)
+        ${CMAKE_CURRENT_SOURCE_DIR}/include/
+        ${CMAKE_CURRENT_SOURCE_DIR}/src/runtime/
```

Comment:
We need a definition of `cldnn::ocl::gpu_buffer` declared in [`plugins/intel_gpu/src/runtime/ocl/ocl_memory.hpp`](https://github.com/dchigarev/openvino/blob/d41b8671cf814f237c2eb3a22c05db5e9efdc4c7/src/plugins/intel_gpu/src/runtime/ocl/ocl_memory.hpp#L29) inside of [`plugins/intel_gpu/src/plugin/ops/mlir_op.cpp`](https://github.com/dchigarev/openvino/blob/d41b8671cf814f237c2eb3a22c05db5e9efdc4c7/src/plugins/intel_gpu/src/plugin/ops/mlir_op.cpp#L60-L61) in order to be able to cast to that type from an abstract `cldnn::memory` and extract an actual `cl_mem` buffer.

Can we do better than this?

Reply:
I think this can be handled w/o exposing OCL internals since you anyway cast cl handles to void* .
So my suggestion is to introduce virtual method `void* get_handle()` for cldnn::stream and other required classes. So the code will look like this:
```
ov::EvaluationContext meta;
meta.insert(ov::intel_gpu::ocl_queue(stream->get_handle()));

std::vector<ov::intel_gpu::gpu_handle_param> events_list;
if (stream.get_queue_type() == cldnn::QueueTypes::out_of_order) {
    events_list.reserve(dependent_events.size() + 1);
    for (auto& ev : dependent_events) {
        events_list.push_back(ev.get_handle());
    }
}
meta.insert(ov::intel_gpu::mlir_meta::wait_list(events_list));

```

Reply:
So I've added the `void* get_handle()` method to `cldnn::stream/buffer/device/event` interfaces. The method is supposed to return a pointer to an opencl handle from C-api (`cl_mem`, `cl_command_queue`, `cl_device_id`, ...) since gc-gpu runtime takes these instead of c++ wrappers. 

However for `cldnn::event` I had to make an exception and [return a pointer to `cl::Event`](https://github.com/dchigarev/openvino/blob/a0981ac515c9ad4306ec52fc73a45094ecea1838/src/plugins/intel_gpu/src/runtime/ocl/ocl_base_event.hpp#L28-L29) instead, since we need a c++ wrapper to [propagate a `cl_event` returned by graph compiler](https://github.com/slyalin/openvino/pull/169#discussion_r1784303743) to openvino's `cldnn::event`.

This exception doesn't look right, I'm wondering whether we should always return pointers to c++ wrappers in `void* get_handle()` and then extract an underlying c-handle manually, or we can live with how it's for now?

Or are there other ways of how we can propagate the returned `cl_event` to `cldnn::event`? Will something like this via `add_event_handler()` work?
```cpp
cldnn::event::ptr ev = stream.create_user_event();
...
cl_event gc_ev = launchGcModuleAndGetEvent();

ev->add_event_handler([](void* ev) {
    CHECK_ERR(clWaitForEvents(1, &reinterpret_cast<cl_event>(ev)));
}, gc_ev);
```